### ISSUE TITLE
EKS: Envelope Encryption of Kubernetes Secrets

### ISSUE THREAD

**Issue by tabern** [aws/containers-roadmap#530]
Kubernetes supports .the ability to enable envelope encryption of Kubernetes secrets using a customer managed key (CMK).

To support a defense-in-depth strategy, we plan to enable the [AWS Encryption Provider](https://github.com/kubernetes-sigs/aws-encryption-provider) on EKS and allow you to provide a [AWS KMS CMK](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#master_keys) that will be used for envelope encryption of Kubernetes secrets on the cluster.

_Note_ This feature will provide an additional security layer for customers using Kubernetes secrets. All secrets and other data stored in the etcd for all EKS clusters is encrypted at rest today. https://docs.aws.amazon.com/en_us/eks/latest/userguide/clusters.html and https://github.com/aws/containers-roadmap/issues/263

**Comment by kekoav** [comment]
@tabern What about supporting an AWS-managed KMS key? We would like envelope encryption, without managing the key ourselves. Is that included in this?

**Comment by tabern** [comment]
@kekoav yes - we're planning to allow two modes. In the first mode, you request encryption but don't provide a key. EKS will create a managed KMS key in your account and associate it to the cluster. In the second mode, you can create a KMS key yourself and then pass the ARN to the cluster.

**Comment by Moulick** [comment]
@tabern Any plans to make k8s secrets be backed or synced with AWS secret manager ?

**Comment by pmnhatdn** [comment]
@tabern I'd like to check if we have an ETA for rolling out this feature as this is affecting our roadmap of managing applications in EKS.

Thanks.

**Comment by bgshacklett** [comment]
@Moulick it looks like that request is being tracked in a separate issue: https://github.com/aws/containers-roadmap/issues/168

**Comment by tabern** [comment]
Hi everyone,

Amazon EKS now supports envelope encryption of K8s secrets using AWS KMS keys!

#### How it works
Envelope encryption for secrets is available for new Amazon EKS clusters running Kubernetes version 1.13 and above. You can setup your own [Customer Master Key (CMK) in KMS](https://docs.aws.amazon.com/kms/latest/developerguide/create-keys.html) and link this key by providing the CMK ARN when you create an EKS cluster. 

When secrets are stored using the Kubernetes secrets API, they are encrypted with a Kubernetes-generated data encryption key, which is then further encrypted using the linked AWS KMS key.

#### More details: 
- [what's new post](https://aws.amazon.com/about-aws/whats-new/2020/03/amazon-eks-adds-envelope-encryption-for-secrets-with-aws-kms/)
- [blog](https://aws.amazon.com/blogs/containers/using-eks-encryption-provider-support-for-defense-in-depth/)
- [documentation](https://docs.aws.amazon.com/eks/latest/userguide/create-cluster.html)

**Comment by khacminh** [comment]
@tabern 

> You can setup your own Customer Master Key (CMK) in KMS and link this key by providing the CMK ARN **when you create an EKS cluster.**

How about the running clusters?

**Comment by tomiszili** [comment]
hi

If there is an option to enable encryption like in S3 without generating any KMS keys by myself, that would be nice.

**Comment by bflad** [comment]
Support for this functionality has been merged into the Terraform AWS Provider and will release with version 2.52.0, later today.

**Comment by jtackaberry** [comment]
Does anyone know why the apiserver locally generates the DEK, as opposed to calling `kms:GenerateDataKey`?

**Comment by jughead** [comment]
Does anyone know what policy and which role to assign the KMS key usage to? When I added the encryption configuration to the same setup process (Terraform), EKS didn't create nodes. Regarding the role, I mean is it for cluster role, for node role?
I tried adding kms:Encrypt, kms:Decrypt with KMS resource to cluster and node roles still the nodes are not created.

**Comment by tabern** [comment]
Update on this feature, on 3/1 we added support for encrypting existing, running clusters. @khacminh 
* [what's new post](https://aws.amazon.com/about-aws/whats-new/2021/03/amazon-eks-supports-adding-kms-envelope-encryption-to-existing-clusters/)

**Comment by chadlwilson** [comment]
Thanks @tabern - I guess https://github.com/aws/containers-roadmap/issues/782 can be closed now in that case?
